### PR TITLE
GraphQl-463: Query category for disabled catalog with products returns the product count

### PR DESCRIPTION
**Pull request by XxXgeoXxX**
### Description (*)
issue https://github.com/magento/graphql-ce/issues/463

#### Steps to reproduce (*)
Create Simple Product and assign it to any category that can be added to menu
Navigate to Catalog -> Categories and edit the category from previous step
Disable the category
Perform the query:
```
{
  category(id: {{ ID }} ) {
    product_count
  }
}
```

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by naydav** [comment]
Has failed internal PAT builds
https://bamboo.corp.magento.com/browse/MCCE23-PAT2482-1

Need to run `jmeter` scenario on the current branch
https://github.com/magento/graphql-ce/blob/2.3-develop/setup/performance-toolkit/benchmark.jmx

**Comment by naydav** [comment]
@XxXgeoXxX 
Could you run jmeter on a local machine?

**Comment by naydav** [comment]
@XxXgeoXxX 
For reproducing this problem you can execute:
```
{
  category(id: 1) {
    name
    id
    level
    description
    path
    path_in_store
    url_key
    url_path
    children {
      id
      description
      default_sort_by
      children {
        id
        description
        level
        children {
          level
          id
          children {
            id
          }
        }
      }
    }
  }
}
```
Result:
```
{
  "errors": [
    {
      "debugMessage": "Notice: Undefined index: is_active in /var/www/graphql.local/docs/ce/app/code/Magento/CatalogGraphQl/Model/Resolver/CategoryTree.php on line 84",
      "message": "Internal server error",
      "category": "internal",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "category"
      ]
    }
  ],
  "data": {
    "category": null
  }
}
```

**Comment by naydav** [comment]
Generally, it is not a good idea to build a tree and just after that throw exception.
We can add `is_active` condition to query for loading tree from db.

**Comment by naydav** [comment]
\Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\CategoryTree::getTree
\Magento\CatalogGraphQl\Model\Category\LevelCalculator::calculate

**Comment by naydav** [comment]
@XxXgeoXxX 
We have static tests failures
```
Magento\Test\Php\LiveCodeTest::testCodeMess
PHP Code Mess has found error(s):
/var/www/html/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php:25	The class CategoryTree has a coupling between objects value of 14. Consider to reduce the number of dependencies under 13.
```
```
FILE: ...logGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 153 | ERROR | [x] Missing short description
 154 | ERROR | [x] There must be exactly one blank line before tags
----------------------------------------------------------------------
```
Could you look at it?
Thanks

**Comment by XxXgeoXxX** [comment]
Short description was added

**Comment by naydav** [comment]
@XxXgeoXxX 
But still have
```
Magento\Test\Php\LiveCodeTest::testCodeMess
PHP Code Mess has found error(s):
/var/www/html/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php:25	The class CategoryTree has a coupling between objects value of 14. Consider to reduce the number of dependencies under 13.
```

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php [hunk 1]
```diff
@@ -144,4 +149,30 @@ private function joinAttributesRecursively(Collection $collection, FieldNode $fi
             $this->joinAttributesRecursively($collection, $node);
         }
     }
+
+    /**
+     * Check if provided category active
+     *
+     * @param int $rootCategoryId
+     * @return bool
+     * @throws \Exception
```
Comment (rogyar) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php]: Please, clarify the exception. It's rather `\Magento\Framework\Exception\LocalizedException` than `\Exception`

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php [hunk 2]
```diff
@@ -84,6 +85,10 @@ public function __construct(
     public function getTree(ResolveInfo $resolveInfo, int $rootCategoryId): \Iterator
     {
         $categoryQuery = $resolveInfo->fieldNodes[0];
+        if ($this->isCategoryActive($rootCategoryId) == false) {
```
Comment (rogyar) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php]: Please, use strict comparison operator in such cases (`===`). Technically, they are faster and if we use type casting (bool) we are sure that we are going to have a boolean type here in any case

**Review by rogyar — CHANGES_REQUESTED** [review]

**Comment by rogyar** [comment]
Hi @XxXgeoXxX. Please, check my review comments.

But prior to fixing the issues, I would like to discuss the approach you've chosen. If we take a look at https://github.com/magento/graphql-ce/blob/5bf736481300478ed6ce2dfd3c3ac5dd6f19a5a9/app/code/Magento/CatalogGraphQl/Model/Resolver/CategoryTree.php#L77 - the logic is built in the way to have the error if categories collection is empty. So, I believe it's a good idea to stick with this approach instead of throwing the same exception twice. All we need to do is to make sure that the category collection is empty on requesting an inactive category.

At the same time, we already have `is_active` check here 

https://github.com/magento/graphql-ce/blob/62fd888833477a6bb35623e333f0e8798afb293f/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php#L104

but looks like something goes wrong here.

So instead of introducing another `is_active` check for the collection, maybe we should fix the existing one and use the existing exception throwing flow?

I might miss something but just want to share my look on this matter. I'm always open to a discussion.

Thank you

**Comment by XxXgeoXxX** [comment]
Hi @rogyar
changes were provided

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php [hunk 3]
```diff
@@ -84,6 +85,10 @@ public function __construct(
     public function getTree(ResolveInfo $resolveInfo, int $rootCategoryId): \Iterator
     {
         $categoryQuery = $resolveInfo->fieldNodes[0];
+        if ($this->isCategoryActive($rootCategoryId) === false) {
```
Comment (TomashKhamlai) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/CategoryTree.php]: Please cover this case:
category with "path": "1/2/3/4/5" has products assigned and is enabled
category with "path": "1/2/3/4" has no products assigned and is enabled
category with "path": "1/2/3" has no products assigned and is disabled

Assert 1: for category 3 product_count throws GraphQlNoSuchEntityException
Assert 2: for category 5 product_count throws GraphQlNoSuchEntityException
Assert 2 will fail.

**Review by naydav — APPROVED** [review]

**Comment by naydav** [comment]
@XxXgeoXxX please look at failed static tests
Looks like we can move `isCategoryActive` method to separate class
Thanks

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=476) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=476) before we can accept your contribution.<br/>**2** out of **15** committers have signed the CLA.<br/><br/>:white_check_mark: naydav<br/>:white_check_mark: lenaorobei<br/>:x: mKovdrysh<br/>:x: magento-devops-community-builds-svc<br/>:x: joanhe<br/>:x: soumyau<br/>:x: magento-cicd2<br/>:x: anthoula<br/>:x: kabothu<br/>:x: cpartica<br/>:x: magento-engcom-team<br/>:x: VladimirZaets<br/>:x: serhii-balko<br/>:x: yelahin-serhiy<br/>:x: diazwatson

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Category/CheckCategoryIsActive.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * Check if category is active.
+     *
+     * @param int $rootCategoryId
+     * @throws GraphQlNoSuchEntityException
+     */
+    public function execute(int $rootCategoryId): void
+    {
+        $collection = $this->collectionFactory->create();
+        $collection->addAttributeToFilter(Category::KEY_IS_ACTIVE, ['eq' => 1])
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Category/CheckCategoryIsActive.php]: Should bereplaced on `\Magento\Catalog\Api\Data\CategoryInterface::KEY_IS_ACTIVE`

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Category/CheckCategoryIsActive.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        $this->metadata = $metadata;
+    }
+
+    /**
+     * Check if category is active.
+     *
+     * @param int $rootCategoryId
+     * @throws GraphQlNoSuchEntityException
+     */
+    public function execute(int $rootCategoryId): void
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Category/CheckCategoryIsActive.php]: The more proper variable name is `categoryId`

**Review by naydav — CHANGES_REQUESTED** [review]

**Review by naydav — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @naydav, thank you for the review.
[ENGCOM-4576](https://jira.corp.magento.com/browse/ENGCOM-4576) has been created to process this Pull Request
:eight_spoked_asterisk: @naydav, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Pull request merged**